### PR TITLE
Luigi move navigation to React

### PR DESCRIPTION
**Pull request by wchrapka**
**Description**

- [x] map static navigation nodes to resources/***/index.js
- [x] create fn to assemble the navigation structure based on re-maped nodes

**Related issue(s)**

See #1939

### CODE REVIEW

File: core-ui/src/resources/Jobs/index.js [hunk 1]
```diff
@@ -1,9 +1,13 @@
 import React from 'react';
-
 import { matchByOwnerReference } from 'shared/utils/helpers';
+import { PredefinedCategories } from 'sidebar/constants';
 
 export const resourceType = 'Jobs';
 export const namespaced = true;
+export const label = 'jobs.title';
+export const apiGroup = '';
+export const apiVersion = 'batch/v1';
```
Comment (Wawrzyn321) [inline core-ui/src/resources/Jobs/index.js:9]:
```suggestion
export const apiGroup = 'batch';
export const apiVersion = 'v1';
```

File: core-ui/src/state/authDataAtom.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { atom, RecoilState } from 'recoil';
+
+type AuthDataState = {
+  'client-certificate-data': string;
+  'client-key-data': string;
+  token?: string;
+} | null;
```
Comment (Wawrzyn321) [inline core-ui/src/state/authDataAtom.ts:7]: It's "either-or" here: we should have either:
- `null` or
- {token: string} or
- {'client-certificate-data': string, 'client-key-data': string}

File: core-ui/src/state/navigation/fetchPermissions.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  const activeNamespaceId = get(activeNamespaceIdState);
+  const postFn = getPostFn(get);
+
+  if (postFn) {
+    const namespaceName = activeNamespaceId ? activeNamespaceId : '*';
+    const path = '/apis/authorization.k8s.io/v1/selfsubjectrulesreviews';
+    const ssrr = {
+      typeMeta: {
+        kind: 'SelfSubjectRulesReview',
+        aPIVersion: 'authorization.k8s.io/v1',
```
Comment (Wawrzyn321) [inline core-ui/src/state/navigation/fetchPermissions.ts:15]: My mistake from made 19 months ago... 🤦 
```suggestion
        APIVersion: 'authorization.k8s.io/v1',
```

File: core-ui/src/state/utils/getFetchFn.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import { createFetchFn } from '../../shared/hooks/BackendAPI/useFetch';
+
+export const getFetchFn = (get: GetRecoilValue) => {
+  const authData = get(authDataState);
+  const cluster = get(clusterState);
+  const clusterConfig = get(clusterConfigState);
+  const ssoData = get(ssoDataState);
+  const { fromConfig } = get(configState) || {};
+
+  if ((authData || ssoData) && cluster && clusterConfig && fromConfig) {
```
Comment (Wawrzyn321) [inline core-ui/src/state/utils/getFetchFn.ts:16]: Only `authData` is required, `ssoData` is optional. Optionally we could check for `ssoData` **only** if SSO_LOGIN feature is enabled?

File: core-ui/src/state/navigation/categories.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+export enum PredefinedCategories {
+  'workloads' = 'Workloads',
+  'discovery-and-network' = 'Discovery-and-network',
+  'istio' = 'Istio',
+  'service-management' = 'Service-management',
+  'storage' = 'Storage',
+  'apps' = 'Apps',
+  'configuration' = 'Configuration',
+  'integration' = 'Integration',
+}
```
Comment (Lyczeq) [inline core-ui/src/state/navigation/categories.ts:12]: Recently I've found an interesting [article](https://frontlive.pl/blog/typescript-enum) (in Polish) about enums. Have you considered replacing them with Objects like in last example in the link above?

File: core-ui/src/shared/hooks/BackendAPI/useFetch.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import { useConfig } from 'shared/contexts/ConfigContext';
+
+export type FetchFn = ({
+  relativeUrl,
+  abortController,
+  init,
+}: {
+  relativeUrl: string;
+  init: any;
+  abortController?: any;
```
Comment (Lyczeq) [inline core-ui/src/shared/hooks/BackendAPI/useFetch.ts:14]: Can we try to add types here?

File: core-ui/src/shared/hooks/BackendAPI/useFetch.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  ssoData: any;
+  fromConfig: any;
+}) => async ({
+  relativeUrl,
+  abortController,
+  init,
+}: {
+  relativeUrl: string;
+  init: any;
+  abortController?: any;
```
Comment (Lyczeq) [inline core-ui/src/shared/hooks/BackendAPI/useFetch.ts:36]: Like above

File: core-ui/src/state/navigation/filters/filterPermittedNodes.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+export const filterPermittedNodes = (
+  nodes: NavNode[],
+  permissionSet: PermissionSet[],
+) => {
+  return nodes.filter(node => isResourcePermitted(node, permissionSet));
+};
+
+const isResourcePermitted = (node: NavNode, permissionSet: PermissionSet[]) => {
+  const resourceGroupAndVersion = `${node.apiGroup}${node.apiGroup ? '/' : ''}${
+    node.apiVersion
```
Comment (Lyczeq) [inline core-ui/src/state/navigation/filters/filterPermittedNodes.ts:14]: is this double `node.apiGroup` necessary?

File: core-ui/src/state/permissionSetsAtom.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    const activeNamespaceId = get(activeNamespaceIdState);
+    const postFn = getPostFn(get);
+
+    if (postFn) {
+      const namespaceName = activeNamespaceId ? activeNamespaceId : '*';
+      const path = '/apis/authorization.k8s.io/v1/selfsubjectrulesreviews';
+      const ssrr = {
+        typeMeta: {
+          kind: 'SelfSubjectRulesReview',
+          aPIVersion: 'authorization.k8s.io/v1',
```
Comment (Lyczeq) [inline core-ui/src/state/permissionSetsAtom.ts:27]:
```suggestion
          apiVersion: 'authorization.k8s.io/v1',
```

File: core-ui/src/state/resourceList/busolaResourcesToNavNodes.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import { NavNode } from '../types';
+
+type PartialNavNode = Partial<NavNode> & Pick<NavNode, 'resourceType'>;
+
+export const busolaResourcesToNavNodes = (resource: PartialNavNode) => {
+  const node: NavNode = {} as NavNode;
+
+  node.category = resource.category || 'temporary';
```
Comment (Lyczeq) [inline core-ui/src/state/resourceList/busolaResourcesToNavNodes.ts:8]: I don't really understand why the optional value is `temporary`

File: core-ui/src/state/types.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  LEGAL_LINKS = 'LEGAL_LINKS',
+  SSO_LOGIN = 'SSO_LOGIN',
+  KUBECONFIG_ID = 'KUBECONFIG_ID',
+  SENTRY = 'SENTRY',
+  OBSERVABILITY = 'OBSERVABILITY',
+  HIDDEN_NAMESPACES = 'HIDDEN_NAMESPACES',
+  VISUAL_RESOURCES = 'VISUAL_RESOURCES',
+  EXTENSIBILITY = 'EXTENSIBILITY',
+  TRACKING = 'TRACKING',
+}
```
Comment (Lyczeq) [inline core-ui/src/state/types.ts:25]: Please, see my comment about enums above :P

**Review by Lyczeq — CHANGES_REQUESTED** [review]

File: core-ui/src/state/navigation/filters/filterPermittedNodes.ts (hunk 8 above)
Comment (wchrapka) [reply core-ui/src/state/navigation/filters/filterPermittedNodes.ts:14]: we want to add / if apiGroup is present. Currently can't think of other solution. any ideas?

File: core-ui/src/state/utils/getFetchFn.ts (hunk 4 above)
Comment (wchrapka) [reply core-ui/src/state/utils/getFetchFn.ts:16]: Let's not go for features check atm, as we have the rerender problem there due to Prometheus

File: core-ui/src/state/navigation/filters/permissions.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    resourceGroupAndVersion === 'v1' ? 'api' : 'apis'
+  }/${resourceGroupAndVersion}/${resourceNamePlural}`.toLowerCase();
+
+  const doesExist = !!resourceIdList.includes(resourcePathId);
+
+  return doesExist;
+};
+
+export const doesUserHavePermission = (
+  permissions = ['get', 'list'],
```
Comment (Wawrzyn321) [inline core-ui/src/state/navigation/filters/permissions.js:20]: I'd go with only "list" for default - only this verb is required for deciding if we want to show or hide the navigation node. The same in `core/src/luigi-config/navigation/permissions.js`:51.

File: core-ui/src/state/navigation/filters/filterScopeNodes.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import { NavNode } from '../../types';
+
+export const filterScopeNodes = (
+  navList: NavNode[],
+  scope: 'cluster' | 'namespace',
+) => {
+  if (!(scope === 'cluster' || scope === 'namespace')) {
+    console.error('Navigation scope is not defined.');
+    return [];
+  }
```
Comment (Wawrzyn321) [inline core-ui/src/state/navigation/filters/filterScopeNodes.ts:10]: I don't think this is needed? TS is already complaining if I use a different string for `scope`.

Comment (Wawrzyn321) [inline core-ui/src/state/navigation/filters/filterScopeNodes.ts:10]:
```suggestion
```
?

**Review by Lyczeq — APPROVED** [review]

**Review by Wawrzyn321 — APPROVED** [review]

**Pull request merged**